GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+        private int bufferContentLength;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        public string ReadMessage()
+        {
+            List<byte> bytes = new List<byte>();
```

Comment:
If you do stick with using a `List`, I would figure out how many bytes we're using for our current set of messages, and preallocate the list with at least that much room, to avoid having to do a reallocation while reading a message.